### PR TITLE
feat(order-client): #MLYSTOR-603 new dates and status for orderclients 

### PR DESCRIPTION
**Pull request by Rseuret**
## Describe your changes
add new dates and get status with isntruction cp
## Issue ticket number and link
https://jira.support-ent.fr/browse/MLYSTOR-603
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [ ] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/resources/public/template/customer/campaign/order/orders-by-equipment.html [hunk 1]
```diff
@@ -19,141 +19,7 @@
             </div>
         </div>
     </div>
-    <div class="cell width-100-40">
```
Comment (Rseuret) [inline src/main/resources/public/template/customer/campaign/order/orders-by-equipment.html:22]: refacto

File: src/main/resources/public/template/customer/campaign/order/orders-by-project.html [hunk 2]
```diff
@@ -42,140 +42,9 @@ <h3 class="cell horizontal-margin">
                         </div>
                     </div>
                 </div>
-                <div class="cell width-100-40">
```
Comment (Rseuret) [inline src/main/resources/public/template/customer/campaign/order/orders-by-project.html:45]: refacto

File: src/main/resources/public/ts/controllers/administration/labelOperation.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import {_, ng, template, moment} from 'entcore';
-import {label, Utils} from "../../model";
+import {Label, Utils} from "../../model";
```
Comment (Rseuret) [inline src/main/resources/public/ts/controllers/administration/labelOperation.ts:2]: refacto label -> Label

File: src/main/resources/public/ts/model/OrderClient.ts [hunk 4]
```diff
@@ -186,8 +239,70 @@ export class OrderClient implements Order  {
...
+        this.tax_amount = orderClientResponse.tax_amount;
+        //supplier
+        this.supplier_name = orderClientResponse.name_supplier;
+        if (orderClientResponse.options && orderClientResponse.options.length > 0 && orderClientResponse.options[0] !== null)
+            this.options = orderClientResponse.options.map(options => {
+                return new OrderOptionClient().build(options);
+            });
+        else
+            this.options = [];
+        //à transformer en contact
```
Comment (Rseuret) [inline src/main/resources/public/ts/model/OrderClient.ts:275]: commentaire fait pour futures évos

File: src/main/resources/public/ts/model/operation.ts [hunk 5]
```diff
@@ -6,15 +6,15 @@ import {Utils} from "./Utils";
 import {Instruction} from "./instruction";
 import {OrderClient} from "./OrderClient";
 import {Structure} from "./Structure";
-import {label} from "./LabelOperation";
+import {Label} from "./LabelOperation";
```
Comment (Rseuret) [inline src/main/resources/public/ts/model/operation.ts:9]: refacto label > Label

File: src/main/resources/public/ts/controllers/main.ts [hunk 6]
```diff
@@ -234,9 +235,10 @@ export const mainController = ng.controller('MainController', ['$scope', 'route'
                 $scope.idIsInteger(idCampaign);
                 if (!$scope.current.structure)
                     await $scope.initStructures();
-                $scope.current.structure
-                    ? await $scope.ordersClient.sync(null, [], [], [], [], [], [], [], idCampaign, $scope.current.structure.id)
-                    : null;
+                //temporaire en attendant de le mettre dans le controller
+                await orderService.sync(idCampaign,$scope.current.structure.id).then( result =>
```
Comment (qpic-cgi) [inline src/main/resources/public/ts/controllers/main.ts:239]: si tu fais un then, tu peux enlever le await !

Comment (qpic-cgi) [inline src/main/resources/public/ts/controllers/main.ts:239]: si tu fais un then, tu peux enlever le await !

File: src/main/resources/public/ts/controllers/viewer/order-list.ts [hunk 7]
```diff
@@ -157,6 +151,7 @@ export const orderPersonnelController = ng.controller('orderPersonnelController'
         };
 
         $scope.deleteProject = async (projects: Projects) => {
+            console.log("deleteProject")
```
Comment (qpic-cgi) [inline src/main/resources/public/ts/controllers/viewer/order-list.ts:154]: supp le console log

File: src/main/resources/public/ts/controllers/viewer/order-list.ts [hunk 8]
```diff
@@ -170,7 +165,10 @@ export const orderPersonnelController = ng.controller('orderPersonnelController'
                     }
                 }
             }
-            await $scope.ordersClient.sync(null, [],[],[],[],[],[], [],$routeParams.idCampaign, $scope.current.structure.id);
+            await orderService.sync($routeParams.idCampaign, $scope.current.structure.id).then( result =>
```
Comment (qpic-cgi) [inline src/main/resources/public/ts/controllers/viewer/order-list.ts:168]: tu peux enlever ton await

File: src/main/java/fr/openent/lystore/service/impl/DefaultProjectService.java [hunk 9]
```diff
@@ -158,8 +159,12 @@ public void handle(Message<JsonObject> event) {
                                 log.error(message);
                                 handler.handle(new Either.Left<>(message));
                             } else {
-                                JsonArray results = event.body().getJsonArray("results");
-                                Double purse_amount = Double.parseDouble(results.getJsonObject(results.size()-1).getJsonArray("results").getJsonArray(0).getString(0));
+                                JsonArray results = event.body().getJsonArray(CommonConstants.RESULTS);
+                                double purse_amount = 0.d;
```
Comment (JinDouang) [inline src/main/java/fr/openent/lystore/service/impl/DefaultProjectService.java:163]: Si possible, faire en camelCase (https://rules.sonarsource.com/java/tag/convention/RSPEC-117)

File: src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    formatDate(date: Date): string {
+        return Utils.formatDate(date);
+    }
+}
+
+function directive(): IDirective {
+    return {
+        replace: true,
```
Comment (JinDouang) [inline src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts:119]: Tu n'es pas obligé de mettre `replace: true`

File: src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            return lang.translate(orderClient.status) + " " + orderClient.operation.label
+        return lang.translate(orderClient.status);
+    }
+
+    displayStatus(): string {
+        return lang.translate(this.orderClient.status)
+    }
+
+    displayInstruction(): string {
+        return " : Rapport " + this.orderClient.operation.instruction.cp_number;
```
Comment (JinDouang) [inline src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts:95]: Possible de le gérer en i18n ?

File: src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    getTooltip(orderClient: OrderClient): string {
+        if (orderClient.operation && orderClient.operation.instruction)
+            return orderClient.operation.instruction.object
+        if (orderClient.operation)
+            return lang.translate(orderClient.status) + " " + orderClient.operation.label
+        return lang.translate(orderClient.status);
+    }
+
+    displayStatus(): string {
+        return lang.translate(this.orderClient.status)
```
Comment (JinDouang) [inline src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts:91]:
```suggestion
        return lang.translate(this.orderClient.status);
```

File: src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+
+    hasAProposalPrice(): boolean {
+        return this.orderClient.price_proposal !== undefined && this.orderClient.price_proposal !== null && this.orderClient.price_proposal !== 0;
+    }
+
+    getTooltip(orderClient: OrderClient): string {
+        if (orderClient.operation && orderClient.operation.instruction)
+            return orderClient.operation.instruction.object
```
Comment (JinDouang) [inline src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts:84]:
```suggestion
            return orderClient.operation.instruction.object;
```

File: src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+        if (orderClient.operation)
+            return lang.translate(orderClient.status) + " " + orderClient.operation.label
```
Comment (JinDouang) [inline src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts:86]:
```suggestion
            return lang.translate(orderClient.status) + " " + orderClient.operation.label;
```

File: src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+    calculatePriceOfOrderClient(orderClient: OrderClient, selectedOptions: boolean, roundNumber: number = 2): number {
+        return (!isNaN(orderClient.calculatePriceTTC(selectedOptions)))
+            ? (roundNumber
+                ? Number.parseFloat(orderClient.calculatePriceTTC(selectedOptions).toFixed(roundNumber))
+                : orderClient.calculatePriceTTC(selectedOptions))
+            : orderClient.calculatePriceTTC(selectedOptions);
+    }
+
+    calculateTotal(orderClient: OrderClient, roundNumber: number): string {
+        let totalPrice = this.calculatePriceOfOrderClient(orderClient, true, roundNumber) * orderClient.amount;
```
Comment (JinDouang) [inline src/main/resources/public/ts/directives/order-client-display-viewer/order-client-display-viewer.directive.ts:73]: typage

File: src/main/resources/public/ts/model/OrderClient.ts [hunk 16]
```diff
@@ -186,8 +238,70 @@ export class OrderClient implements Order  {
...
+        this.price = orderClientResponse.price;
+        this.price_proposal = orderClientResponse.price_proposal;
+        this.rank = orderClientResponse.rank;
+        this.status = orderClientResponse.status;
+        this.summary = orderClientResponse.summary;
+        this.tax_amount = orderClientResponse.tax_amount;
+        //supplier
+        this.supplier_name = orderClientResponse.name_supplier;
+        if (orderClientResponse.options && orderClientResponse.options.length > 0 && orderClientResponse.options[0] !== null)
+            this.options = orderClientResponse.options.map(options => {
```
Comment (JinDouang) [inline src/main/resources/public/ts/model/OrderClient.ts:269]: typage `options ` dans ta map

tu pourrais aussi le faire en une ligne : 

```suggestion
            this.options = orderClientResponse.options.map((options: IOrderClientOptionResponse) => new OrderOptionClient().build(options));
```

File: src/main/resources/public/ts/model/OrderClient.ts [hunk 17]
```diff
@@ -523,31 +609,33 @@ export class OrdersClient extends Selection<OrderClient> {
...
-    amount: number;
-    required: boolean;
-    id_order_client_equipment: number;
-    selected: boolean;
+    build(orderClientsResponse: IOrderClientResponse[]): OrdersClient {
+        this.all = orderClientsResponse.map((orderClientResponse: IOrderClientResponse) => {
+            return new OrderClient().build(orderClientResponse);
+        });
+        this.projects = new Selection<Project>([]);
+        this.all.forEach(order =>{
```
Comment (JinDouang) [inline src/main/resources/public/ts/model/OrderClient.ts:618]: typage

File: src/main/resources/public/ts/model/OrderClient.ts [hunk 18]
```diff
@@ -523,31 +609,33 @@ export class OrdersClient extends Selection<OrderClient> {
...
+            if(this.projects.filter(projectFiltered => projectFiltered.id === order.project.id).length === 0)
+                return this.projects.all.push(order.project);
+        })
```
Comment (JinDouang) [inline src/main/resources/public/ts/model/OrderClient.ts:621]:
```suggestion
        });
```

File: src/main/resources/public/ts/model/OrderClient.ts [hunk 19]
```diff
@@ -523,31 +609,33 @@ export class OrdersClient extends Selection<OrderClient> {
...
-    required: boolean;
-    id_order_client_equipment: number;
-    selected: boolean;
+    build(orderClientsResponse: IOrderClientResponse[]): OrdersClient {
+        this.all = orderClientsResponse.map((orderClientResponse: IOrderClientResponse) => {
+            return new OrderClient().build(orderClientResponse);
+        });
+        this.projects = new Selection<Project>([]);
+        this.all.forEach(order =>{
+            if(this.projects.filter(projectFiltered => projectFiltered.id === order.project.id).length === 0)
```
Comment (JinDouang) [inline src/main/resources/public/ts/model/OrderClient.ts:619]: typage

File: src/main/resources/public/ts/services/app/index.ts [hunk 20]
```diff
@@ -0,0 +1,2 @@
+export * from './admin/index'
+export * from './viewer/index'
```
Comment (JinDouang) [inline src/main/resources/public/ts/services/app/index.ts:2]:
```suggestion
export * from './admin/index';
export * from './viewer/index';
```

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/lystore/service/impl/DefaultOrderService.java [hunk 21]
```diff
@@ -138,7 +146,26 @@ public void listOrder(Integer idCampaign, String idStructure, Handler<Either<Str
...
+                            elem.put(LystoreBDD.OPTIONS, new JsonArray(elem.getString(LystoreBDD.OPTIONS)));
+                            elem.put(LystoreBDD.TITLE, new JsonObject(elem.getString(LystoreBDD.TITLE)));
+                            elem.put(LystoreBDD.PRICE, OrderUtils.safeGetDouble(elem, LystoreBDD.PRICE));
+                            elem.put(LystoreBDD.PRICE_PROPOSAL, OrderUtils.safeGetDouble(elem, LystoreBDD.PRICE_PROPOSAL));
+                            elem.put(LystoreBDD.TAX_AMOUNT, OrderUtils.safeGetDouble(elem, LystoreBDD.TAX_AMOUNT));
+                            return  elem;
+                        })
+                        .collect(Collectors.toList()))));
+            } else {
+                handler.handle(new Either.Left<>(event.left().getValue()));
```
Comment (JinDouang) [inline src/main/java/fr/openent/lystore/service/impl/DefaultOrderService.java:166]: log erreur (avec string.format etc...) si possible!

**Review by JinDouang — CHANGES_REQUESTED** [review]

**Review by JinDouang — APPROVED** [review]

**Pull request merged**